src/main/scala/com/fulcrumgenomics/bam/ErrorRateByReadPosition.scala

```diff
@@ -0,0 +1,280 @@
+
+@clp(description =
+  """
+    |Calculates the error rate by read position on mapped BAMs.
+    |
+    |If a reference is given, the sort order must be coordinate.  If no reference is given, the MD tag must be present
+    |for every mapped read and computed correctly, and no sort order is required.
+    |
+    |The output metrics text file will have the extension 'error_rate_by_read_position.txt'.  The output file will
+    |contain a row per read and per position in that read.  The 'read_number' will be 0 for fragments, 1 for the first
+    |end of a pair, and 2 for the second end of a pair. Both the error rate and number of bases observed will be
+    |provided.  Soft-clipped bases, and unmapped, secondary, supplementary, and vendor failing reads are ignored.
+    |Optionally, duplicates are ignored.
+  """,
+  group = ClpGroups.SamOrBam)
+class ErrorRateByReadPosition
+( @arg(flag="i", doc="Input BAM file.") val input: PathToBam,
+  @arg(flag="o", doc="Output metrics prefix. If not given, will use the input BAM basename as a prefix.")
+  val output: Option[PathPrefix] = None,
+  @arg(flag="r", doc="Reference fasta. If not given, will use the MD tag.") val ref: Option[PathToFasta] = None,
+  @arg(flag="d", doc="Include duplicate reads, otherwise ignore.") val duplicates: Boolean = false,
+  @arg(flag="I", doc="Count insertions as errors, otherwise as matches.") val includeInsertions: Boolean = false,
+  @arg(flag="m", doc="The minimum mapping quality for a read to be included.") val minMappingQuality: Int = 0,
+  @arg(flag="q", doc="The minimum base quality for a base to be included.") val minBaseQuality: Int = 0
+) extends FgBioTool with LazyLogging {
+  import scala.collection.JavaConversions.asScalaIterator
+
+  Io.assertReadable(input)
+  ref.foreach(Io.assertReadable)
+  output.foreach(out => Io.assertCanWriteFile(out))
+
+  def execute(): Unit = {
+    val progress = new ProgressLogger(logger, verb="read", unit=1e6.toInt)
+    val in       = SamReaderFactory.make.open(input)
+    val iterator = in.iterator()
+    val counter  = ErrorRateByReadPositionCounter(ref=ref, includeInsertions=includeInsertions, minBaseQuality=minBaseQuality)
+
+    if (ref.exists(r => in.getFileHeader.getSortOrder != SortOrder.coordinate)) {
+      throw new IllegalArgumentException("Input BAM must be coordinate sorted when the reference is given.")
+    }
+
+    iterator.filterNot { rec =>
+      rec.getReadUnmappedFlag || rec.isSecondaryOrSupplementary ||
+        (!duplicates && rec.getDuplicateReadFlag) || rec.getReadFailsVendorQualityCheckFlag ||
+        (rec.getMappingQuality < minMappingQuality)
+    }
+    .foreach { rec =>
+      counter.accept(rec)
+      progress.record(rec)
+    }
+
+    val path = PathUtil.pathTo(output.getOrElse(PathUtil.removeExtension(input)) + ErrorRateByReadPositionMetric.FileExtension)
+    Metric.write[ErrorRateByReadPositionMetric](metrics=counter.metrics, path=path)
```

Comment:
Is the type annotation needed here?  If `counter.metrics` has a declared return type I would think it could be inferred.